samples/nrf9160/serial_lte_modem/src/slm_at_gps.c

```diff
@@ -52,6 +57,156 @@ void rsp_send(const u8_t *str, size_t len);
 /* global variable defined in different files */
 extern struct at_param_list at_param_list;
 extern char rsp_buf[CONFIG_AT_CMD_RESPONSE_MAX_LEN];
+extern struct k_work_q slm_work_q;
+
+#ifdef CONFIG_SUPL_CLIENT_LIB
+static int supl_fd;
+int connect_supl_server(void)
+{
+	int ret;
+	struct sockaddr_in server;
+	struct addrinfo *res;
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM,
+		.ai_protocol = IPPROTO_TCP,
+		/* Either a valid,
+		 * NULL-terminated access point name or NULL.
+		 */
+		.ai_canonname = NULL,
+	};
+
+	ret = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+	if (ret < 0) {
+		LOG_ERR("socket() error: %d", -errno);
+		return -errno;
+	}
+
+	supl_fd = ret;
+	ret = getaddrinfo(CONFIG_SLM_SUPL_SERVER, NULL, &hints, &res);
+	if (ret || res == NULL) {
+		LOG_ERR("getaddrinfo() error: %d", -errno);
+		close(supl_fd);
+		return ret;
+	}
+	server.sin_family = AF_INET;
+	server.sin_port = htons(CONFIG_SLM_SUPL_PORT);
+	server.sin_addr.s_addr =
+		((struct sockaddr_in *)res->ai_addr)->sin_addr.s_addr;
+	ret = connect(supl_fd, (struct sockaddr *)&server,
+		sizeof(struct sockaddr_in));
+	if (ret) {
+		LOG_ERR("connect() error: %d", -errno);
+		close(supl_fd);
+		freeaddrinfo(res);
+		return -errno;
+	}
+
+	freeaddrinfo(res);
+	return 0;
+}
+
+ssize_t supl_write(const void *p_buff, size_t nbytes, void *user_data)
+{
+	ARG_UNUSED(user_data);
+	return send(supl_fd, p_buff, nbytes, 0);
+}
+
+int supl_logger(int level, const char *fmt, ...)
+{
+	char buffer[256] = { 0 };
+	va_list args;
+
+	va_start(args, fmt);
+	int ret = vsnprintk(buffer, sizeof(buffer), fmt, args);
+
+	va_end(args);
+
+	if (ret < 0) {
+		LOG_ERR("[SUPL] Encoding error");
+		return ret;
+	} else if ((size_t)ret >= sizeof(buffer)) {
+		LOG_WRN("[SUPL] Too long message");
+	}
+
+	LOG_DBG("[SUPL] %s", log_strdup(buffer));
+
+	return ret;
+}
+
+ssize_t supl_read(void *p_buff, size_t nbytes, void *user_data)
+{
+	ARG_UNUSED(user_data);
+	ssize_t rc = recv(supl_fd, p_buff, nbytes, 0);
+
+	if (rc < 0 && (errno == ETIMEDOUT)) {
+		return 0;
+	}
+
+	return rc;
+}
+
+int supl_inject(void *agps, size_t agps_size, nrf_gnss_agps_data_type_t type,
+	     void *user_data)
+{
+	int ret;
+	ARG_UNUSED(user_data);
+
+	ret = nrf_sendto(client.sock, agps, agps_size, 0, &type, sizeof(type));
```

Comment:
`socket` and `nrf_socket` APIs should not be mixed (it will not work, underlying socket FD used by nrf API might be different than the one used by Zephyr).
If it's not possible to use `socket` API for all operations, I'd switch to nrf for now.

Reply:
Nevermind, we don't mix APIs on a single socket here so it's fine.

Reply:
Thanks